mantidproject/mantid
# Attempt to fix conda system tests

`master_systemtests-conda` build has been failing due to a zero length vector normalization error in `SolidAngle` during `ILLDirectGeometryReductionTest.IN4` system test. This PR attempts to fix to the issue by catching a possible exception when normalizing a vector in `CSGObject::isOnSize()`. Unfortunately, things got a bit out of hand and this PR contains a hefty amount of refactoring, especially regarding the new free function `normalize()` introduced here.

**To test:**

Code review. I do not know how to reproduce the issue outside the specific RHEL7 master pipeline. Our standard system test build runs the offending test without problems. The ultimate test would be to check `master_systemtests-conda` after this has been merged to master.

Fixes #25373. 

*This does not require release notes* because it is an internal change fixing a system test on an unsupported system (Ubuntu 14.04)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
